### ISSUE TITLE
Speeding up Axes3D.plot_surface 4-8x

### ISSUE THREAD

**Issue by apaszke** [matplotlib/matplotlib#16659]
Hi! I've recently had to produce a lot of 3D surface plots and have noticed that the implementation of [`Axes3D.plot_surface`](https://matplotlib.org/mpl_toolkits/mplot3d/tutorial.html#mpl_toolkits.mplot3d.Axes3D.plot_surface) is heavily suboptimal for my use case. I think that the implementation has been made so generic is that for some reason `plot_surface` has to deal with subsampling the input internally, which [may produce polygons with boundaries of different lengths](https://github.com/matplotlib/matplotlib/blob/3205ff797038fb4662b843faceeaaec1f63acef9/lib/mpl_toolkits/mplot3d/axes3d.py#L1570-L1571). In my case, the plot is an output of a simulation, so I want to display it in its full resolution (i.e. `rstride == cstride == 1`), meaning that all polygons are actually squares. This has allowed me to store all their data in a NumPy arrays instead of super long Python lists, and I've been able to **speed up the whole rendering process by 4-8x** by monkey patching the following methods:
* `Axes3D.plot_surface`
* `Poly3DCollection.do_3d_projection`
* `Poly3DCollection.get_vector`
* `PolyCollection.set_verts`

**The question I have for you is: would you be interested in having me upstream those changes?** The only concern I would have in mind is that it might split some code paths in two (one dealing with lists, the other with NumPy arrays), potentially making future maintenance more difficult. At the same time, if someone wants to draw plots without subsampling the input grid, or with strides are divisors of the grid size, then 4x speedups can be quite noticeable. In my case I've been able to reduce the rendering time for a single animation (100 plots with 200x200 points in a grid) from 2min 21s to 18s, or a single higher-resolution frame (1 plot with 800x800 points in a grid) from 1min 13s to 11s, and those do make a difference in your workflow. I am perfectly willing to wait a few seconds, but definitely not minutes.

I've also added a `Poly3DCollection.update_surface` method, which allowed me to reuse the existing plot object for even faster rendering of a series of plots.

As mentioned before I already have all of the code necessary, but right monkey patching means it will likely get out of sync with the future versions, and everyone else misses out on the work I've already put in. Let me know how would you like to proceed with that.

By the way I've been interested in pushing the speedups even further, but it looks like those improvements would likely require storing collections of paths describing sets of homogeneous polygons in arrays instead of Python lists and adapting the renderers. That seems like much more work, but I'd be very excited if matplotlib could produce those plots in at least near real-time speeds.

**Matplotlib version**
  * Operating system: Linux
  * Matplotlib version: 3.1.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): Agg
  * Python version: 3.7

**Comment by jklymak** [comment]
My guess is people would be interested, and `mplot3d` is not exactly over maintained at the moment, so having someone enthusiastic about improving it would be great.   The issue will probably be if your changes are back compatible or not.  Did you run your changes on our tests and are there visible discrepancies?  ping @WeatherGod  and @eric-wieser in particular for guidance.

**Comment by apaszke** [comment]
I've been trying to be very careful when modifying the code and have my own tests to make sure I'm not messing up my visualizations, but in some cases I have pre-specialized certain code paths just because it was meant to be a quick hack so far. If you'd be interested in upstreaming this, then I would treat my code as a template for the changes I would have to make, but I'd redo all of it to make sure I haven't broken anything else along the way. Right now I'm mostly trying to figure out if this extra effort is even worth it, because I don't want to do that if it's going to get rejected.

**Comment by eric-wieser** [comment]
For anyone reading along `plot_surface` is defined here: https://github.com/matplotlib/matplotlib/blob/3205ff797038fb4662b843faceeaaec1f63acef9/lib/mpl_toolkits/mplot3d/axes3d.py#L1432-L1599

**Comment by tacaswell** [comment]
I am nominally :+1: on adding a fast-path, but conditional on how much spaghetti it adds.

**Comment by apaszke** [comment]
It shouldn't be too bad, because NumPy arrays behave pretty much the same as lists of tuples. Seems like there's a fair chance it might be useful, so I'll try to start submitting them in small chunks soon.

**Comment by eric-wieser** [comment]
If you've already written the monkey-patching, it would be great if you could post it in a gist so that we can evaluate the amount of spaghetti.

I think a fast-path for `rstride == cstride == 1` would be quite reasonable, since you could then avoid the call to `cbook._array_perimeter`.

**Comment by apaszke** [comment]
Yes, exactly! IIRC the biggest improvements were:
- Avoiding `cbook._array_perimeter`
- Vectorizing all transforms on the Z values
- Removing a lot of list comprehensions/transposes in `Poly3DCollection.do_3d_projection`

**Comment by apaszke** [comment]
[Here's the gist](https://gist.github.com/apaszke/2b797a614315365fe00aac01af1662db)

**Comment by eric-wieser** [comment]
Just checking you're aware, but if you set `rstride = cstride = 10`, then each facet on the surface contains 40 points, not 4 - a curve is plotted along the perimeter of the subsampling.

I suppose the question is whether the overhead is caused by the subsampling itself, or by the resulting non-uniform list of polygons.

Perhaps  `Poly3DCollection` can do something clever like store a list of polygon sizes, and pad the coordinate arrays.

**Comment by apaszke** [comment]
Yes, I know that in those cases they will have more vertices, but it might still let me remove a lot of list comprehensions and turn those into array ops. Also, another good point you make, but I think would be a slightly bigger effort, would be to ditch lists and use masked arrays.

**Comment by eric-wieser** [comment]
You might want to look at #11577

**Comment by apaszke** [comment]
@eric-wieser Interesting! I think it's a step in a good direction and may let you vectorize some of it, but it still does include some long loops/comprehensions ([one example here](https://github.com/matplotlib/matplotlib/pull/11577/files#diff-fc54f86f287bf203dfa9e78102e2678bR609)) that you can avoid when all the faces have the same number of vertices.

**Comment by apaszke** [comment]
Is there any plan for the PR you've linked? If there's anyone who's going to pick it up soon then we should synchronize, because it will lead to nasty conflicts.

**Comment by apaszke** [comment]
Thought it might be reasonable to link this to the PRs I've submitted so far: #16675, #16689, #16688, #16699, #16793.

Assuming all of those are merged the overall time for plotting a 400x400 surface on my old Macbook drops from 15.85s to 0.86s (18x speedup!!).

Another improvement that could be made to drop the time even further to around 0.34s (based on my hacky implementation) would be to avoid storing homogeneous paths as separate objects in a Python list, and keep both vertices and codes vectorized in a special collection. To fully unlock its potential I would also have to modify the C++ bits for the AGG backend (mostly both classes from `src/py_adaptors.h`), which might make them a bit more low level, and could introduce a separate optimized code path. Would that be an acceptable change, or is that going too far? I'm happy to hack on this in the next few days, but would like to have some confirmation that there is a chance that it will get merged.

**Comment by anntzer** [comment]
I can't guarantee it will be merged, but I would be happy to at least try to review a PR using a better structure of homogenous path collections.

**Comment by scottshambaugh** [comment]
Hey @apaszke, I am taking a look into 3D plotting performance right now, and see your open PRs https://github.com/matplotlib/matplotlib/pull/16688, https://github.com/matplotlib/matplotlib/pull/16699, https://github.com/matplotlib/matplotlib/pull/16793 as good improvements. Are you interested in finishing those up? If not I can bring them through to the finish line.
